x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportDeleteDataFrameTransformAction.java

```diff
@@ -5,93 +5,72 @@
+        PersistentTasksCustomMetaData pTasksMeta = state.getMetaData().custom(PersistentTasksCustomMetaData.TYPE);
+        if (pTasksMeta != null && pTasksMeta.getTask(request.getId()) != null) {
+            listener.onFailure(new ElasticsearchStatusException("Cannot delete data frame [" + request.getId() +
+                    "] as the task is running. Stop the task first", RestStatus.CONFLICT));
         } else {
-            listener.onFailure(new IllegalStateException("Could not delete transform [" + request.getId() + "] because "
-                    + "indexer state is [" + state + "].  Transform must be [" + IndexerState.STOPPED + "] before deletion."));
+            // Task is not runing, delete the configuration document
+            transformsConfigManager.deleteTransform(request.getId(), ActionListener.wrap(
+                    r -> listener.onResponse(new AcknowledgedResponse(r)),
+                    listener::onFailure));
         }
     }
 
     @Override
-    protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
-        final ClusterState state = clusterService.state();
-        final DiscoveryNodes nodes = state.nodes();
-        if (nodes.isLocalNodeElectedMaster()) {
-            PersistentTasksCustomMetaData pTasksMeta = state.getMetaData().custom(PersistentTasksCustomMetaData.TYPE);
-            if (pTasksMeta != null && pTasksMeta.getTask(request.getId()) != null) {
-                super.doExecute(task, request, listener);
-            } else {
-                // we couldn't find the transform in the persistent task CS, but maybe the transform exists in the configuration index,
-                // if so delete the orphaned document and do not throw (for the normal case we want to stop the task first,
-                // than delete the configuration document if and only if the data frame transform is in stopped state)
-                transformsConfigManager.deleteTransform(request.getId(), ActionListener.wrap(r -> {
-                    listener.onResponse(new Response(true));
-                    return;
-                }, listener::onFailure));
-            }
-        } else {
-            // Delegates DeleteTransform to elected master node, so it becomes the coordinating node.
-            // Non-master nodes may have a stale cluster state that shows transforms which are cancelled
-            // on the master, which makes testing difficult.
-            if (nodes.getMasterNode() == null) {
-                listener.onFailure(new MasterNotDiscoveredException("no known master nodes"));
-            } else {
-                transportService.sendRequest(nodes.getMasterNode(), actionName, request,
-                        new ActionListenerResponseHandler<>(listener, Response::new));
-            }
-        }
+    protected ClusterBlockException checkBlock(Request request, ClusterState state) {
+        return null;
```

Comment:
Wouldn't it be a good idea to make sure that the `metadata` read information is not dirty before executing this task?
```
return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_READ);
```